File: Builder/Python/Ex1.py
# PALETA DE CORES
ui = {
    "reset": "\033[0m",
    "bold": "\033[1m",
    "pink": "\033[35m",
    "purple": "\033[38;5;54m",
    "white": "\033[97m"
}

# Classe base de Pessoa
class Pessoa:
    def __init__(self, nome, cargo, matricula, hora_entrada, hora_saida):
        self.nome = nome
        self.cargo = cargo
        self.matricula = matricula
        self.hora_entrada = hora_entrada
        self.hora_saida = hora_saida

    def mostrar_detalhes(self):
        print(
            f"{ui['purple']}│ "
            f"{ui['white']}{self.nome.ljust(12)}"
            f"{ui['purple']} │ "
            f"{ui['white']}{self.cargo.ljust(14)}"
            f"{ui['purple']} │ "
            f"{ui['white']}{str(self.matricula).ljust(10)}"
            f"{ui['purple']} │ "
            f"{ui['white']}{self.hora_entrada.ljust(8)}"
            f"{ui['purple']} │ "
            f"{ui['white']}{self.hora_saida.ljust(8)}"
            f"{ui['purple']} │"
            f"{ui['reset']}"
        )

# Builder base
class PessoaBuilder:
    def __init__(self):
        self.nome = ""
        self.cargo = ""
        self.matricula = ""
        self.hora_entrada = ""
        self.hora_saida = ""

    def set_nome(self, nome):
        self.nome = nome
        return self

    def set_matricula(self, matricula):
        self.matricula = matricula
        return self

    def set_hora_entrada(self, hora):
        self.hora_entrada = hora
        return self

    def set_hora_saida(self, hora):
        self.hora_saida = hora
        return self

    def construir(self):
        return Pessoa(
            self.nome,
            self.cargo,
            self.matricula,
            self.hora_entrada,
            self.hora_saida
        )

# Builders específicos
class AlunoBuilder(PessoaBuilder):
    def __init__(self):
        super().__init__()
        self.cargo = "Aluno"

class ProfessorBuilder(PessoaBuilder):
    def __init__(self):
        super().__init__()
        self.cargo = "Professor"

class AdministradorBuilder(PessoaBuilder):
    def __init__(self):
        super().__init__()
        self.cargo = "Administrador"

class VisitanteBuilder(PessoaBuilder):
    def __init__(self):
        super().__init__()
        self.cargo = "Visitante"

# Uso
aluno = (AlunoBuilder()
         .set_nome("Carlos")
         .set_matricula(1234)
         .set_hora_entrada("08:00")
         .set_hora_saida("12:00")
         .construir())

professor = (ProfessorBuilder()
             .set_nome("Ana")
             .set_matricula(5678)
             .set_hora_entrada("09:00")
             .set_hora_saida("17:00")
             .construir())

administrador = (AdministradorBuilder()
                 .set_nome("João")
                 .set_matricula(4321)
                 .set_hora_entrada("07:00")
                 .set_hora_saida("15:00")
                 .construir())

visitante = (VisitanteBuilder()
             .set_nome("Maria")
             .set_matricula("-")
             .set_hora_entrada("10:00")
             .set_hora_saida("12:00")
             .construir())

print(f"{ui['purple']}{ui['bold']}\n════════════════════════ SISTEMA DE PESSOAS ════════════════════════{ui['reset']}")
print(f"{ui['purple']}┌──────────────┬────────────────┬────────────┬──────────┬──────────┐{ui['reset']}")
print(f"{ui['purple']}│ "
      f"{ui['pink']}{'Nome'.ljust(12)}"
      f"{ui['purple']} │ "
      f"{ui['pink']}{'Cargo'.ljust(14)}"
      f"{ui['purple']} │ "
      f"{ui['pink']}{'Matrícula'.ljust(10)}"
      f"{ui['purple']} │ "
      f"{ui['pink']}{'Entrada'.ljust(8)}"
      f"{ui['purple']} │ "
      f"{ui['pink']}{'Saída'.ljust(8)}"
      f"{ui['purple']} │{ui['reset']}")
print(f"{ui['purple']}├──────────────┼────────────────┼────────────┼──────────┼──────────┤{ui['reset']}")

aluno.mostrar_detalhes()
professor.mostrar_detalhes()
administrador.mostrar_detalhes()
visitante.mostrar_detalhes()

print(f"{ui['purple']}└──────────────┴────────────────┴────────────┴──────────┴──────────┘{ui['reset']}")

File: Builder/JavaScript/Ex1.js
// PALETA DE CORES
const ui = {
    reset: "\x1b[0m",
    bold: "\x1b[1m",
    pink: "\x1b[35m",         
    purple: "\x1b[38;5;54m",  
    white: "\x1b[97m"
};

// Classe base de Pessoa
class Pessoa {
    constructor(nome, cargo, matricula, horaEntrada, horaSaida) {
        this.nome = nome;
        this.cargo = cargo;
        this.matricula = matricula;
        this.horaEntrada = horaEntrada;
        this.horaSaida = horaSaida;
    }

    mostrarDetalhes() {
        console.log(
            ui.purple + "│ " +
            ui.white + this.nome.padEnd(12) +
            ui.purple + " │ " +
            ui.white + this.cargo.padEnd(14) +
            ui.purple + " │ " +
            ui.white + this.matricula.toString().padEnd(10) +
            ui.purple + " │ " +
            ui.white + this.horaEntrada.padEnd(8) +
            ui.purple + " │ " +
            ui.white + this.horaSaida.padEnd(8) +
            ui.purple + " │" +
            ui.reset
        );
    }
}

// Builder base
class PessoaBuilder {
    constructor() {
        this.nome = "";
        this.cargo = "";
        this.matricula = "";
        this.horaEntrada = "";
        this.horaSaida = "";
    }

    setNome(nome) {
        this.nome = nome;
        return this;
    }

    setMatricula(matricula) {
        this.matricula = matricula;
        return this;
    }

    setHoraEntrada(hora) {
        this.horaEntrada = hora;
        return this;
    }

    setHoraSaida(hora) {
        this.horaSaida = hora;
        return this;
    }

    construir() {
        return new Pessoa(
            this.nome,
            this.cargo,
            this.matricula,
            this.horaEntrada,
            this.horaSaida
        );
    }
}

// Builders específicos
class AlunoBuilder extends PessoaBuilder { 
    constructor() { 
        super(); 
        this.cargo = "Aluno"; 
    } 
}

class ProfessorBuilder extends PessoaBuilder { 
    constructor() {
        super(); 
        this.cargo = "Professor"; 
    } 
}

class AdministradorBuilder extends PessoaBuilder { 
    constructor() { 
        super(); 
        this.cargo = "Administrador"; 
    } 
}
class VisitanteBuilder extends PessoaBuilder { 
    constructor() { 
        super(); 
        this.cargo = "Visitante"; 
    } 
}

// Criando pessoas
console.log(ui.purple + ui.bold + "\n════════════════════════ SISTEMA DE PESSOAS ════════════════════════" + ui.reset);

console.log(ui.purple + "┌──────────────┬────────────────┬────────────┬──────────┬──────────┐" + ui.reset);
console.log(
    ui.purple + "│ " +
    ui.pink + "Nome".padEnd(12) +
    ui.purple + " │ " +
    ui.pink + "Cargo".padEnd(14) +
    ui.purple + " │ " +
    ui.pink + "Matrícula".padEnd(10) +
    ui.purple + " │ " +
    ui.pink + "Entrada".padEnd(8) +
    ui.purple + " │ " +
    ui.pink + "Saída".padEnd(8) +
    ui.purple + " │" +
    ui.reset
);
console.log(ui.purple + "├──────────────┼────────────────┼────────────┼──────────┼──────────┤" + ui.reset);

const aluno = new AlunoBuilder()
    .setNome("Carlos")
    .setMatricula(1234)
    .setHoraEntrada("08:00")
    .setHoraSaida("12:00")
    .construir();

const professor = new ProfessorBuilder()
    .setNome("Ana")
    .setMatricula(5678)
    .setHoraEntrada("09:00")
    .setHoraSaida("17:00")
    .construir();

const administrador = new AdministradorBuilder()
    .setNome("João")
    .setMatricula(4321)
    .setHoraEntrada("07:00")
    .setHoraSaida("15:00")
    .construir();

const visitante = new VisitanteBuilder()
    .setNome("Maria")
    .setMatricula("-")
    .setHoraEntrada("10:00")
    .setHoraSaida("12:00")
    .construir();

aluno.mostrarDetalhes();
professor.mostrarDetalhes();
administrador.mostrarDetalhes();
visitante.mostrarDetalhes();

console.log(ui.purple + "└──────────────┴────────────────┴────────────┴──────────┴──────────┘" + ui.reset);